Source/WebCore/dom/Element.cpp

```diff
@@ -2436,18 +2444,40 @@ std::optional<Vector<Ref<Element>>> Element::elementsArrayForAttributeInternal(c
             return std::nullopt;
 
         SpaceSplitString ids(getAttribute(attr), SpaceSplitString::ShouldFoldCase::No);
-        return compactMap(ids, [&](auto& id) {
+        elements = compactMap(ids, [&](auto& id) {
             return getElementByIdIncludingDisconnected(*this, id);
         });
     }
 
-    return std::nullopt;
+    if (!elements.has_value())
+        return std::nullopt;
+
+    if (document().settings().shadowRootReferenceTargetEnabled()) {
+        elements = compactMap(elements.value(), [&](Ref<Element>& element) -> std::optional<Ref<Element>> {
+            if (RefPtr deepReferenceTarget = element->deepShadowRootReferenceTargetOrSelf())
+                return *deepReferenceTarget;
+            return std::nullopt;
+        });
+    }
+
+    return elements;
 }
 
 std::optional<Vector<Ref<Element>>> Element::getElementsArrayAttributeForBindings(const QualifiedName& attributeName) const
 {
     ASSERT(isElementsArrayReflectionAttribute(attributeName));
-    return elementsArrayForAttributeInternal(attributeName);
+    std::optional<Vector<Ref<Element>>> elements = elementsArrayForAttributeInternal(attributeName);
+
+    if (!elements)
+        return std::nullopt;
+
+    if (document().settings().shadowRootReferenceTargetEnabled()) {
+        return compactMap(elements.value(), [&](Ref<Element>& element) -> std::optional<Ref<Element>> {
```

Comment:
I don't think we need to use compactMap here since we're not filtering content.  Just use a regular map instead and you can get rid of the return type: `-> std::optional<Ref<Element>>`.